Azure Data Explorer dashboard reports - README (#1929)
microsoft/azure_arc: docs/azure_jumpstart_ag/contoso_supermarket/adx_dashboards/_index.md

---
type: docs
weight: 100
toc_hide: true
---

# Contoso Supermarket data pipeline and reporting across cloud and edge - Store orders

> If you are on this page, by now you may have the familiarity of Contoso Supermarket, its objectives, and goals. As part of the series of various technology stack demonstration scenarios, this scenario covers the data pipeline and reporting across various Contoso Supermarket stores and services to showcase how the Point of Sale (PoS) order data flows to Azure Data Explorer (ADX) database and generate near realtime reports. Contoso management can leverage these reports to adjust their inventory and supply chain based on the product demand from customer orders across stores, day, week, month, and year. This helps optimize Contoso resources, store capacity, and saves significant cost and at the same time improves customer satisfaction and trust.

Reviewer: Always start with a problem statement. "Contoso need to achieve X..."

One of Contoso's biggest objectives is how to use the data coming from their stores and visualize it for business intelligence by leveraging the power of the cloud.

In this scenario, Contoso wants to use their data pipeline so customer orders placed in the Point of Sale (PoS) application on the various supermarket stores, flow to [Azure Data Explorer (ADX)](https://learn.microsoft.com/azure/data-explorer/data-explorer-overview) database and generate near real-time reports. By doing so, Contoso management can leverage these reports to adjust their inventory and supply chain based on the product demand from customer orders across multiple filters - stores, day, week, month, and year. This helps optimize Contoso resources, stores supplies, saves significant costs and at the same time improves customer satisfaction and trust.

Author: Updated as per the comments

## Architecture

> Below is an architecture diagram that shows how the data flows from PoS application and all the way into Azure Data Explorer (ADX) database to generate near realtime reports of orders received and processed across various store locations. This architecture includes a local PostgreSQL database running at the edge location of the Contoso store, Cosmos DB and ADX cluster in Azure cloud, and a Cloud sync service that moves orders data from edge location to Cosmos DB in the cloud. Please refer to steps 1 through 6 in the architecture diagram below to understand how the orders data flows to ADX database.

Reviewer: ```suggestion
Below is an architecture diagram that shows how the data flows from the PoS application and into the ADX database to generate near real-time reports of orders received and processed across various supermarket store locations. This architecture includes a local PostgreSQL database running at the edge in the store, [Azure Cosmos DB](https://learn.microsoft.com/azure/cosmos-db/introduction) and ADX cluster in Azure cloud, and a Cloud Sync service that moves orders data from edge location to Cosmos DB in the cloud.
```

![Data pipeline architecture diagram](./img/contoso_supermarket_pos_service_architecture.png)

## PoS dashboard reports

> Contoso supports dashboard reports for Point of Sale (PoS) application and Internet of Things (IoT) environment sensor analytics and monitoring. These reports are created in Azure Data Explorer (ADX) to allow users to view dashboards reports. These reports are generated based on live data received from PoS application and IoT environment sensors into ADX database using data integration.

Reviewer: ```suggestion
Contoso supports dashboard reports for the PoS application and Internet of Things (IoT) environment sensor analytics and monitoring. These reports are created in ADX to allow users to view dashboards reports. These reports are generated based on live data received from the PoS application and IoT environment sensors into the ADX database using data integration.
```

> ## Import dashboards

Reviewer: ```suggestion
## Manually import dashboards
```

> > **Note:** You can skip this step if you are using azd for deployment as these reports are imported for you during the deployment.

Reviewer: ```suggestion
> **Note:** If you used the Azure Developer CLI (azd) method to deploy the Contoso Supermarket scenario, you can skip this step as these reports are automatically imported for you during the deployment.
```

> In order to view PoS Order dashboard reports you need to import these dashboards into ADX. You should have completed deployment of Jumptstart Agora in your environment and logon script is completed after you first login to Client VM _Ag-VM-Client_. Follow the step below to import dashboards once all the pre-requisites are completed.

Reviewer: ```suggestion
Follow the below steps In order to view the PoS Orders dashboard reports you will need to import these into ADX. 
```

> - On the Client VM, open Windows Explorer and navigate to folder _C:\Ag\adx_dashboards_ folder. This folder contains two ADX dashboard report JSON files (_adx-dashboard-iotsensor-payload.json_ and _adx-dashboard-orders-payload.json_) with ADX name and URI updated when the logon script is completed.

Reviewer: ```suggestion
- On the Client VM, open Windows Explorer and navigate to folder _C:\Ag\adx_dashboards_ folder. This folder contains two ADX dashboard report JSON files (_adx-dashboard-iotsensor-payload.json_ and _adx-dashboard-orders-payload.json_) with the ADX name and URI updated when the deployment PowerShell logon script is completed.
```

  ![Locate dashboard report template files](./img/adx_dashboard_report_files.png)

> - Copy these ADX dashboard report JSON files on your computer in a temporary folder to import into ADX dashboards. Alternatively you can login to ADX Dashboards directly on the Client VM. Your Azure AD tenant may have conditional access policies enabled and might prevent login to ADX Dashboards from the Client VM as these VM is not managed by your organization.

Author: Updated as per your feedback

> - On your computer or Client VM open Edge browser and login to [ADX Dashboards](https://dataexplorer.azure.com/). Use the same user account that you deployed Jumpstart Agora in your subscription. Failure to use same account will prevent access to the ADX Orders database to generate reports.

Reviewer: ```suggestion
- On your computer or Client VM open Edge browser and log in to [ADX Dashboards](https://dataexplorer.azure.com/). Use the same user account that you deployed Jumpstart Agora in your subscription. Failure to use the same account will prevent access to the ADX Orders database to generate reports.
```

What if they don't use Edge in their local env? you need to generalize

- On the Client VM **Ag-VM-Client**, locate Data Emulator icon on the desktop.

  ![Locate Data Emulator on the desktop](./img/locate_data_emulator_desktop.png)

- Double click on the Data Emulator desktop icon to launch executable and generate sample data. Confirm by entering **Yes** or **Y** to start generating data, entering No or N will exit the tool. This tool generates data for the last 30 days. Say No or N to prevent regenerating sample data if it is generated earlier. You can still generate additional sample data. Please note, there might be duplicate key errors and might fail to generate data in subsequent attempts.

  ![Confirm sample data generation](./img/confirm_sample_data_generation.png)

  ![Generating sample data](./img/sample_data_generation.png)

- From ADX open PoS Orders report to see newly simulated orders data. Allow some time to propagate data into ADX database using integrated data pipeline.

  ![PoS Orders with simulated data](./img/adx_posorders_with_simulated_data.png)

> - By default report displays for the last one hour data. To change report time range, chose time range as last 30 days from the report to refresh and display for the selected time range.

Author: Rephrased sentence

  ![PoS Orders select time range](./img/adx_orders_report_select_timerange.png)

  ![PoS Orders with simulated data for selected time range](./img/adx_posorders_with_simulated_data_selected_timerange.png)

## Generate orders from Contoso Supermarket store applications

- On the Agora client VM **Ag-VM-Client**, open Edge browser. From the favorites bar review bookmarks created for PoS applications for different stores and environments.

  ![PoS app bookmarks](./img/pos_app_edge_bookmarks.png)

- From the bookmarks expand POS -> Chicago and select POS Chicago - Customer application.

  ![Select POS Chicago application](./img/pos_app_edge_select_pos_chicago_customer.png)

- Randomly add few items to the cart.

  ![Select PoS Chicago application](./img/chicago_pos_app_customer.png)

- Click on Cart, review items in the cart, and click Place Order.

  ![Select PoS Chicago application](./img/pos_chicago_customer_place_order.png)

> - Place multiple orders by adding random items to the cart.

Author: Rephrased sentence.

## Manually import dashboards

> **Note:** If you used the [Azure Developer CLI (azd) method](https://github.com/microsoft/azure_arc/blob/jumpstart_ag/docs/azure_jumpstart_ag/contoso_supermarket/deployment/_index.md#deployment-via-azure-developer-cli-experimental) to deploy the Contoso Supermarket scenario, you may skip this section as these reports are automatically imported for you during the deployment.

Follow the below steps in order to view the PoS Orders dashboard reports you will need to import these into ADX.

- On the Client VM, open Windows Explorer and navigate to folder _C:\Ag\adx_dashboards_ folder. This folder contains two ADX dashboard report JSON files (_adx-dashboard-iotsensor-payload.json_ and _adx-dashboard-orders-payload.json_) with the ADX name and URI updated when the deployment PowerShell logon script is completed.

  ![Locate dashboard report template files](./img/adx_dashboard_report_files.png)

> - Copy these ADX dashboards report JSON files on your computer in a temporary folder to import into ADX dashboards. Alternatively, you can log in to ADX Dashboards directly on the Client VM.

Reviewer: ```suggestion
- Copy these ADX dashboards report JSON files on your local machine in a temporary folder to import into ADX dashboards. Alternatively, you can log in to ADX Dashboards directly on the Client VM.
```

  > **Note:** Depending the account being used to log in to ADX portal, Azure AD tenant of that account may have conditional access policies enabled and might prevent log in to ADX Dashboards from the Client VM as this VM is not managed by your organization.

> - On your computer open browser of your choice or on the Client VM open Edge browser and log in to [ADX Dashboards](https://dataexplorer.azure.com/). Use the same user account that you deployed Jumpstart Agora in your subscription. Failure to use the same account will prevent access to the ADX Orders database to generate reports.

Reviewer: ```suggestion
- On your local machine open the browser of your choice OR on the Client VM open the Edge browser and log in to [ADX Dashboards](https://dataexplorer.azure.com/). Use the same user account that you deployed Jumpstart Agora in your subscription. Failure to use the same account will prevent access to the ADX Orders database to generate reports.
```

> - Once you log in to ADX dashboards, click on Dashboards in the left navigation to import PoS Orders dashboard report.

Reviewer: ```suggestion
- Once you are logged in to ADX dashboards, click on Dashboards in the left navigation to import the PoS Orders dashboard report.
```

  ![Navigate to ADX dashboard reports](./img/adx_view_dashboards.png)

> - Select _Import dashboard from file_ to select previously copied file from the Client VM or _C:\Ag\adx_dashboards_ folder on the Client VM.

Reviewer: ```suggestion
- Select _Import dashboard from file_ to select previously copied file from the Client VM to your local machine or the _C:\Ag\adx_dashboards_ folder on the Client VM.
```

  ![Select import dashboard file](./img/adx_import_dashboard_file.png)

> - Choose _adx-dashboard-orders-payload.json_ file to import.

Reviewer: ```suggestion
- Choose to import the _adx-dashboard-orders-payload.json_ file.
```

  ![Choose dashboard report JSON file to import](./img/adx_select_dashboard_file.png)

> - Confirm dashboard report name, accept the suggested name or chose your own name and click _Create_.

Reviewer: ```suggestion
- Confirm the dashboard report name, accept the suggested name (or choose your own), and click Create.
```

- PoS Orders dashboard report is configured to display data from the _"Last 1 hour"_ by default. To view all the simulated orders data, change report time range to _"Last 30 days"_ as shown in the picture below. Dashboard report will refresh data and display reports for the selected time range.

  ![PoS Orders select time range](./img/adx_orders_report_select_timerange.png)

  ![PoS Orders with simulated data for selected time range](./img/adx_posorders_with_simulated_data_selected_timerange.png)

### Generate orders from Contoso Supermarket store applications

- On the Agora client VM, open Edge browser. From the favorites bar review bookmarks created for PoS applications for different stores and environments.

  ![PoS app bookmarks](./img/pos_app_edge_bookmarks.png)

- From the bookmarks expand POS -> Chicago and select "POS Chicago - Customer".

  ![Select POS Chicago application](./img/pos_app_edge_select_pos_chicago_customer.png)

- Randomly add a few items to the cart.

  ![Select PoS Chicago application](./img/chicago_pos_app_customer.png)

- Click on Cart, review items, and click Place Order.

  ![Select PoS Chicago application](./img/pos_chicago_customer_place_order.png)

> - Place additional orders from the same store by repeating the above steps. Try adding random items to each order to simulate orders from different customers of the store.
> 
> - To view the orders placed form the Contoso Chicago store, go to ADX portal, under Dashboards open PoS Orders report and chose time range as _"Last 30 minutes"_.
> 
>   > **Note:** Some times it takes time to flow orders data in to the ADX database. Wait for few minutes and refresh report to see orders data.

Reviewer: ```suggestion
- In the ADX Portal, under Dashboards, open the PoS Orders report and set the time range for "_Last 30 minutes_", and refresh the report to see data.

> **NOTE:** As the Cloud Sync service performs the sync in the backend, it might take a few minutes for orders to show up in ADX.
```

> - To view the orders placed form the Contoso Chicago store, go to ADX portal, under Dashboards open PoS Orders report and chose time range as _"Last 30 minutes"_.
> 
>   > **Note:** Some times it takes time to flow orders data in to the ADX database. Wait for few minutes and refresh report to see orders data.

Reviewer: ```suggestion
- In the ADX Portal, under Dashboards, open the PoS Orders report and set the time range for "_Last 30 minutes_", and refresh the report to see data.

> **NOTE:** As the Cloud Sync service performs the sync in the backend, it might take a few minutes for orders to show up in ADX.
```
